gh-116126: Implement PEP 696

JelleZijlstra opened the pull request:
This implements both the grammar and compiler changes and changes to `typing.py` behavior (the latter copied from typing-extensions).

<!-- gh-issue-number: gh-116126 -->
* Issue: gh-116126
<!-- /gh-issue-number -->

JelleZijlstra committed:
gh-116126: Grammar changes for PEP 696

JelleZijlstra committed:
Fix test_ast

JelleZijlstra committed:
new tests

JelleZijlstra committed:
fix test_unparse

JelleZijlstra committed:
possible solution

JelleZijlstra committed:
Can't use the name either

JelleZijlstra committed:
blurb

JelleZijlstra committed:
should not have pushed that

JelleZijlstra committed:
maybe this will compile

JelleZijlstra committed:
Feature version support

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pep696v2

JelleZijlstra committed:
default_ -> default_value

Suggestion by Brandt.

JelleZijlstra committed:
ast docs

JelleZijlstra committed:
more docs

AlexWaygood commented on Doc/library/typing.rst:
```diff
@@ -1616,7 +1616,7 @@ without the dedicated syntax, as documented below.
 .. _typevar:
 
-.. class:: TypeVar(name, *constraints, bound=None, covariant=False, contravariant=False, infer_variance=False)
+.. class:: TypeVar(name, *constraints, bound=None, covariant=False, contravariant=False, infer_variance=False, default=None)
```
I don't think the signature of `TypeVar` itself should have `default=None`, because otherwise it's impossible to distinguish a TypeVar without a default from one that actually defaults to having type `None` (https://github.com/python/typing_extensions/pull/85)

JelleZijlstra replied:
Good point, that's actually currently incorrect in the runtime code: `TypeVar("T")` and `TypeVar("T", default=None)` cannot be distinguished. Do you know how we should present this in the docs? Maybe this:
```suggestion
.. class:: TypeVar(name, *constraints, bound=None, covariant=False, contravariant=False, infer_variance=False, default=<marker>)
```

AlexWaygood replied:
not sure. Some other options:

```suggestion
.. class:: TypeVar(name, *constraints, bound=None, covariant=False, contravariant=False, infer_variance=False, default=<unrepresentable>)
```

"overload-style" -- this is already done elsewhere, e.g. for the constructor at https://docs.python.org/3/library/stdtypes.html#str -- but annoyingly verbose here...

```suggestion
.. class:: TypeVar(name, *constraints, bound=None, covariant=False, contravariant=False, infer_variance=False)
           TypeVar(name, *constraints, default, bound=None, covariant=False, contravariant=False, infer_variance=False)
```

"man-page" style (does it make sense to mix "man-page" syntax with the Python syntax for defaults that the following parameters use?):

```suggestion
.. class:: TypeVar(name, *constraints[, default], bound=None, covariant=False, contravariant=False, infer_variance=False)
```

JelleZijlstra replied:
Thanks, I like the `"<unrepresentable>"` variant best. The overload style is too verbose and mixing the bracket style and normal defaults looks weird.

JelleZijlstra committed:
Fix typevarobject.c

JelleZijlstra committed:
default=<unrepresentable>

JelleZijlstra committed:
NoneType is immortal

AlexWaygood commented on Lib/test/test_typing.py:
```diff
@@ -580,6 +580,69 @@ def test_constructor(self):
+                self.assertEqual(z.__default__, typevar.__default__)
+
+
+
```
too many blank lines

```suggestion
```

JelleZijlstra replied:
Fixed in one of the commits I just pushed

AlexWaygood replied:
yeah, I saw :-)

JelleZijlstra committed:
Runtime implementation

JelleZijlstra committed:
What's New

JelleZijlstra committed:
Document None/NoneType weirdness

JelleZijlstra committed:
fix doctest

carljm commented on Lib/typing.py:
```diff
@@ -280,6 +282,12 @@ def _collect_parameters(args):
+                elif default_encountered:
+                    raise TypeError(f'Type parameter {t!r} without a default'
+                                    ' follows type parameter with a default')
+
```
I don't see any added tests for this behavior?

JelleZijlstra replied:
This one fails if I remove this block:

```
FAIL: test_erroneous_generic (test.test_typing.TypeVarLikeDefaultsTests.test_erroneous_generic)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jelle/py/cpython/Lib/test/test_typing.py", line 633, in test_erroneous_generic
    with self.assertRaises(TypeError):
        Test = Generic[DefaultStrT, T]
AssertionError: TypeError not raised
```

carljm commented on Lib/typing.py:
```diff
@@ -297,8 +305,27 @@ def _check_generic(cls, parameters, elen):
-                        f" actual {alen}, expected {elen}")
+        expect_val = elen
+        if hasattr(cls, "__parameters__"):
+            parameters = [p for p in cls.__parameters__ if get_origin(p) is not Unpack]
```
It's not immediately clear to me what the implications are of excluding `Unpack` here, or why it's correct to do so. Are there tests related to this? (I don't see `Unpack` usage added to `test_typing.py`.) It seems like this is worth tests and a comment.

It also seems a bit strange to me in terms of division of responsibilities and maintainability that `_check_generic`, which is called in various cases, would have special-case code in it that appears to be only relevant to the callsite in `_generic_class_getitem`, and uses `hasattr` to detect whether it's relevant. Would it be reasonable to extract this code to a separate function and extend the signature of `_check_generic` to e.g. take both a minimum and maximum number of expected parameters? The minimum could even be passed as an optional callback, if there's a concern about inefficiency of unnecessarily checking defaults when all parameters have been provided.

JelleZijlstra replied:
Thanks, I realized the implementation for TypeVarTuple defaults is wrong in various places; I need to make a number of changes.

JelleZijlstra replied:
I pushed some changes to make TypeVarTuple work; I need to add more test cases though. I also posted on https://discuss.python.org/t/pep-696-type-defaults-for-typevarlikes/22569/36 to get more opinions on what the grammar should look like.

(I wasn't too keen on adding support for TypeVarTuple defaults in the first place. I doubt that part of the feature is going to get much use.)

AlexWaygood commented on Lib/typing.py:
```diff
@@ -297,8 +305,27 @@ def _check_generic(cls, parameters, elen):
+            if alen < elen:
+                # since we validate TypeVarLike default in _collect_type_vars
+                # or _collect_parameters we can safely check parameters[alen]
+                if getattr(parameters[alen], '__default__', None) is not None:
```
```suggestion
                if hasattr(parameters[alen], '__default__'):
```

AlexWaygood replied:
Oh, I suppose this is deliberately written the way it is currently so that TypeVarLikes with defaults can be differentiated from those without them, since a `__default__` attribute set to `None` has special meaning in this context? If so, I think adding a comment might be warranted, as that definitely wasn't obvious to me at first glance :-)

JelleZijlstra replied:
I copied this from typing-extensions without thinking too hard; I wonder if it's fine to do:
```suggestion
                if parameters[alen].__default__ is not None:
```
Because in CPython we can assume the TypeVars also come from us and therefore must have a `__default__` attribute.

JelleZijlstra replied:
Applied my suggestion as it passes tests.

AlexWaygood commented on Lib/typing.py:
```diff
@@ -297,8 +305,27 @@ def _check_generic(cls, parameters, elen):
+                    return
+
+                num_default_tv = sum(getattr(p, '__default__', None)
+                                        is not None for p in parameters)
```
Slightly weird indentation here

JelleZijlstra replied:
Removed the `getattr` and now it fits on one line.

AlexWaygood commented on Misc/NEWS.d/next/Core and Builtins/2024-02-29-18-55-45.gh-issue-116129.wsFnIq.rst:
This news entry looks out of date; you're doing more than just grammar updates here now

JelleZijlstra replied:
Updated.

AlexWaygood commented on Objects/typevarobject.c:
Maybe the docstrings for TypeVar, ParamSpec and TypeVarTuple should also be updated to mention defaults?

JelleZijlstra replied:
Updated all three.

AlexWaygood reviewed (commented):
The docs look great. I'll dig into the `typing.py` changes more tomorrow (it's late here right now). I don't feel qualified to review the other parts of the PR :)

carljm reviewed (commented):
Looks reasonable to me. (I'm lacking expertise in the parser and grammar.)

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pep696v2

JelleZijlstra committed:
Feedback from Alex

JelleZijlstra committed:
Partial work

JelleZijlstra committed:
Allow * in TypeVarTuple default

JelleZijlstra committed:
Test and fix grammar

JelleZijlstra committed:
Fix TypeVarTuple substitution

JelleZijlstra committed:
Two more cases

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pep696v2

JelleZijlstra committed:
Fix TypeVar substitution

JelleZijlstra committed:
Fix ParamSpec

JelleZijlstra committed:
Prohibit default after TypeVarTuple

JelleZijlstra committed:
Wrap new syntax in run_code()

JelleZijlstra committed:
Silence Ruff

erlend-aasland commented on Doc/library/ast.rst:
```diff
@@ -1848,19 +1860,25 @@ aliases.
 .. class:: TypeVarTuple(name)
 
    A :class:`typing.TypeVarTuple`. ``name`` is the name of the type variable tuple.
+   ``default_value`` is the default value, if any.
```
You need to update the `.. class` def to include the *default\_value* parameter.

erlend-aasland replied:
(Also, we should create a follow-up PR to make sure all parameter names are marked up according to the devguide.)

JelleZijlstra replied:
Filed https://github.com/python/cpython/issues/118181; I'll fix the ones touched by this PR though.

picnixz commented on Doc/library/typing.rst:
```diff
@@ -1616,7 +1616,7 @@ without the dedicated syntax, as documented below.
 .. _typevar:
 
-.. class:: TypeVar(name, *constraints, bound=None, covariant=False, contravariant=False, infer_variance=False)
+.. class:: TypeVar(name, *constraints, bound=None, covariant=False, contravariant=False, infer_variance=False, default=<unrepresentable>)
```
May I suggest having two constructors? that way, the distinction between specifying a default and not specificying it would be perhaps clearer? 

```suggestion
.. class:: TypeVar(name, *constraints, bound=None, covariant=False, contravariant=False, infer_variance=False)
		   TypeVar(name, *constraints, default, bound=None, covariant=False, contravariant=False, infer_variance=False)
```

AlexWaygood replied:
Previously discussed at https://github.com/python/cpython/pull/116129#discussion_r1559924536

picnixz replied:
Ah thank you Alex (I should have read the previous comments more carefully!)

JelleZijlstra replied:
I feel that would be less clear with such a long signature; it makes people search to spot the difference between the signatures.

picnixz commented on Doc/library/typing.rst:
```diff
@@ -1754,15 +1754,27 @@ without the dedicated syntax, as documented below.
 
+   .. attribute:: __default__
+
+      The default value of the type variable, if any. This attribute is ``None``
```
I think an example should be definitely helpful because people might really be confused between `TypeVar('T')` and `TypeVar('T', default=None)`.

AlexWaygood replied:
Yeah, I'm somewhat concerned about the subtlety here. I just posted an alternative suggestion at https://discuss.python.org/t/pep-696-type-defaults-for-typevarlikes/22569/38, though it may be that it's too close to the beta deadline for us to consider making design changes to the PEP now.

AlexWaygood replied:
(FWIW, I don't think that design question that I raised on Discourse should block the initial implementation here being merged.)

picnixz replied:
I don't think it should be a blocker, but I support your suggestion on discourse. 

Now, whatever is decided, the representation shouldn't change too many times afterwards. I think a similar issue (but I don't know whether it was formalized or not) is about the `<not given>` for enumerations, where we had some different representations across versions. This could definitely be some use cases for [PEP 661](https://peps.python.org/pep-0661/) (but I don't know if this one is actively needed). And depending on how you backports things, having a public value known to be the sentinel is better when you test representation outputs (the sentinel for enumerations is private).

JelleZijlstra replied:
For BC reasons we probably can't afford to change this behavior later. I guess we can merge this version first and tweak the implementation during the beta phase, though.

AlexWaygood replied:
> For BC reasons we probably can't afford to change this behavior later. I guess we can merge this version first and tweak the implementation during the beta phase, though.

yes, that was what I was thinking

picnixz replied:
Maybe it's obvious but what does "BC" mean in this context?

AlexWaygood replied:
Backwards Compatibility

picnixz commented on Objects/typevarobject.c:
```diff
@@ -23,6 +23,8 @@ typedef struct {
     PyObject *evaluate_bound;
     PyObject *constraints;
     PyObject *evaluate_constraints;
+    PyObject *default_;
```
I don't know the exact policy for the C API, but if you were to use `default_value` in the AST layout, I think it would make sense to also call it `default_value` here and everywhere else. At least, it captures the semantic of being a default *value* and not just a default *type*.

JelleZijlstra replied:
Sure, renaming this. Note this field is not part of the C API in any way though; it's not even exposed outside of the file.

picnixz replied:
Ah yes, sorry. I never contributed (nor looked) at the C code, so I'm not very familiar with what is being exposed or not (let's say that the "P" stands for private :smile:).

picnixz commented on Lib/test/test_typing.py:
```diff
@@ -580,6 +580,148 @@ def test_constructor(self):
         self.assertIs(T.__bound__, None)
 
 
+class TypeParameterDefaultsTests(BaseTestCase):
```
You might want to test type parameters without defaults following parameters with defaults.

JelleZijlstra replied:
There is a test for that, `test_erroneous_generic`.

picnixz replied:
Ah, actually I was wondering whether you wanted to test a bit more precisely your error messages (I suspected that this one should do the job but it's a huge assertRaises so I wondered whether it covered everything).

JelleZijlstra replied:
Good call. I noticed one of the later tests was also broken.

AlexWaygood commented on Lib/test/test_type_params.py:
```diff
@@ -1148,3 +1160,103 @@ class Inner[U](T):
+
+class DefaultsTest(unittest.TestCase):
+    def test_defaults_on_func(self):
+        ns = run_code("""
+            def func[T=int, **U=float, *V=None]():
+                pass
+        """)
```
An alternative solution is just to add this file to this list here of files excluded from Ruff for now:

https://github.com/python/cpython/blob/eb927e9fc823de9539fcb82c9ea9d055462eb04a/Lib/test/.ruff.toml#L2-L20

We could probably remove it again fairly soon; the necessary updates to Ruff's parser probably won't take _too_ long? But I think this is also fine, and it's nice to keep the list of files excluded from Ruff as small as possible.

JelleZijlstra replied:
I considered that but it's actually sort of good to put all the new syntax in strings, so if I mess up the compiler, it's easier to isolate the failure.

AlexWaygood commented on Lib/typing.py:
```diff
@@ -280,25 +284,51 @@ def _collect_parameters(args):
+            expect_val = elen
+
+        raise TypeError(f"Too {'many' if alen > elen else 'few'} arguments"
+                        f" for {cls}; actual {alen}, expected {expect_val}")
```
I'd find this function so much more readable if we renamed `alen` and `elen` to `actual_len` and `expected_len`, respectively. It felt like it took me way too long to refresh my memory about what this function did :-(

picnixz replied:
Either that or `len/maxlen` (because, when you have default parameters, this doesn't mean that you need to specify everything).

JelleZijlstra replied:
I'm going with Alex's suggestion and I'll also rename "parameters" to "arguments" (parameters are the things in the definition, arguments are what you pass when you do a specialization).

AlexWaygood commented on Lib/typing.py:
```diff
@@ -280,25 +284,51 @@ def _collect_parameters(args):
+        # deal with defaults
+        if alen < elen:
+            # since we validate defaults in _collect_type_vars
+            # or _collect_parameters, we can safely check parameters[alen]
```
This comment's a bit opaque right now:
- There is no `_collect_type_vars()` function (did you mean `_collect_parameters`?
- What do you mean by "we validate defaults"?
- What do you mean by "safely check"?

Maybe something like this?

```suggestion
            # A this point, we've already validated in `_collect_parameters()`
            # that any parameters with defaults appear at the end of the
            # parameters list. As such, we can guarantee here that...
```

[I'm not actually sure what we can guarantee as a result of this, you tell me 😄]

picnixz replied:
> What do you mean by "we validate defaults"?

I think it's about validating the order (required, then optional with defaults).

> What do you mean by "safely check"?

The order being checked, you know that you have `Params = [T1, ..., Tk, D1, ..., Dk]` with `Ti` required and `Di` optional. Thus, `Params[alen].__default__` being not None means that there is a default value. So, everything that comes after has also a default. So you don't have "unfilled" values.

JelleZijlstra replied:
I think the comment actually doesn't add much. At this point, we know that `actual_len < len(cls.__parameters__)`, so obviously `cls.__parameters__[actual_len]` is safe; that doesn't need references to other functions.

JelleZijlstra replied:
Oh actually, the non-obvious part is that we *only* have to check the one at actual_len. If that one has a default, it implies that all later params also have a default.

AlexWaygood replied:
Yeah, my main point is that I find the current comment confusing; I don't feel too strongly about whether it's clarified or removed

JelleZijlstra replied:
Pushed a reworded version

JelleZijlstra committed:
Fix parameter markup

JelleZijlstra committed:
default_ -> default_value

JelleZijlstra committed:
Fix errors

JelleZijlstra committed:
arguments, not parameters

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pep696v2

AlexWaygood commented on Doc/library/ast.rst:
```diff
@@ -1823,19 +1825,29 @@ aliases.
 
-.. class:: ParamSpec(name)
+   .. versionchanged:: 3.13
+      Added ``default_value``.
```
```suggestion
      Added the *default_value* parameter.
```

AlexWaygood commented on Doc/library/ast.rst:
```diff
@@ -1874,6 +1892,9 @@ aliases.
    .. versionadded:: 3.12
 
+   .. versionchanged:: 3.13
+      Added ``default_value``.
```
```suggestion
      Added the *default_value* parameter.
```

AlexWaygood commented on Doc/library/ast.rst:
```diff
@@ -1848,19 +1860,25 @@ aliases.
 
-.. class:: TypeVarTuple(name)
+   .. versionchanged:: 3.13
+      Added ``default_value``.
```
```suggestion
      Added the *default_value* parameter.
```

AlexWaygood commented on Lib/typing.py:
```diff
@@ -280,25 +284,52 @@ def _collect_parameters(args):
+        if actual_len < expected_len:
+            # If the parameter at actual_len has a default, then all parameters
+            # after it must also have one, because we validated as much in
+            # _collect_parameters().
```
Thanks, this is much more comprehensible now. I'd still go for something slightly longer, though; it took me a little bit to wrap my head around this:

```suggestion
            # If the parameter at index `actual_len` in the parameters list
            # has a default, then all parameters after it must also have
            # one, because we validated as much in _collect_parameters().
            # That means that no error needs to be raised here, despite
            # the number of arguments being passed not matching the number
            # of parameters: all parameters that aren't explicitly
            # specialized in this call are parameters with default values.
```

AlexWaygood commented on Lib/typing.py:
```diff
@@ -1089,11 +1120,15 @@ def _typevartuple_prepare_subst(self, alias, args):
+    if left == alen - right and self.__default__ is not None:
+        replacement = _unpack_args([self.__default__])
+    else:
+        replacement = args[left: alen - right]
```
blegh, maybe I'll do a PR going through the whole file replacing all the `alen` and `elen` variables after this is merged 😛

AlexWaygood replied:
Thoughts on making `_unpack_args` accept a variadic number of arguments? Constructing a list here just to pass it to `_unpack_args` feels kinda strange. Something like this?

```diff
--- a/Lib/typing.py
+++ b/Lib/typing.py
@@ -331,7 +331,7 @@ def _check_generic_specialization(cls, arguments):
                         f" for {cls}; actual {actual_len}, expected {expect_val}")
 
 
-def _unpack_args(args):
+def _unpack_args(*args):
     newargs = []
     for arg in args:
         subargs = getattr(arg, '__typing_unpacked_tuple_args__', None)
@@ -1121,7 +1121,7 @@ def _typevartuple_prepare_subst(self, alias, args):
         raise TypeError(f"Too few arguments for {alias};"
                         f" actual {alen}, expected at least {plen-1}")
     if left == alen - right and self.__default__ is not None:
-        replacement = _unpack_args([self.__default__])
+        replacement = _unpack_args(self.__default__)
     else:
         replacement = args[left: alen - right]
 
@@ -1427,8 +1427,7 @@ def __getitem__(self, args):
         # Preprocess `args`.
         if not isinstance(args, tuple):
             args = (args,)
-        args = tuple(_type_convert(p) for p in args)
-        args = _unpack_args(args)
+        args = _unpack_args(*(_type_convert(p) for p in args))
         new_args = self._determine_new_args(args)
         r = self.copy_with(new_args)
         return r
         ```

JelleZijlstra replied:
Sure, done.

picnixz commented on Lib/typing.py:
```diff
@@ -1552,7 +1589,12 @@ def __getitem__(self, params):
-        _check_generic(self, params, self._nparams)
+        if not self._nparams:
+            raise TypeError(f"{self} is not a generic class")
+        alen = len(params)
```
```suggestion
        actual_len = len(params)
```

Let's use the same variable name here as well!

picnixz commented on Python/symtable.c:
```diff
@@ -2285,28 +2303,39 @@ symtable_visit_type_param(struct symtable *st, type_param_ty tp)
-                VISIT_QUIT(st, 0);
+
+        // We must use a different key for the bound and default. The obvious choice would be to
+        // use the .bound and .default_ pointers, but that fails when the expression immediately
```
```suggestion
        // use the .bound and .default_value pointers, but that fails when the expression immediately
```

JelleZijlstra commented on Lib/typing.py:
```diff
@@ -1552,7 +1589,12 @@ def __getitem__(self, params):
+        alen = len(params)
+        if alen != self._nparams:
+            raise TypeError(f"Too {'many' if alen > self._nparams else 'few'} arguments for {self};"
+                            f" actual {alen}, expected {self._nparams}")
```
```suggestion
        if actual_len != self._nparams:
            raise TypeError(f"Too {'many' if actual_len > self._nparams else 'few'} arguments for {self};"
                            f" actual {actual_len}, expected {self._nparams}")
```

JelleZijlstra committed:
Apply suggestions from code review

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>
Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

JelleZijlstra committed:
regen clinic

JelleZijlstra committed:
variadic args for _unpack_args

AlexWaygood commented on Lib/typing.py:
```diff
@@ -1552,7 +1592,12 @@ def __getitem__(self, params):
+        actual_len = len(params)
+        if actual_len != self._nparams:
+            raise TypeError(f"Too {'many' if actual_len > self._nparams else 'few'} arguments for {self};"
+                            f" actual {actual_len}, expected {self._nparams}")
```
small micro-optimisation:

```suggestion
        actual_len = len(params)
        if actual_len != self._nparams:
            if not self._nparams:
                raise TypeError(f"{self} is not a generic class")
            raise TypeError(f"Too {'many' if actual_len > self._nparams else 'few'} arguments for {self};"
                            f" actual {actual_len}, expected {self._nparams}")
```

JelleZijlstra committed:
Update Lib/typing.py

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>

AlexWaygood commented on Doc/whatsnew/3.13.rst:
```diff
@@ -89,6 +89,8 @@ Interpreter improvements:
 
 New typing features:
 
+* :pep:`696`: Type parameters (:data:`typing.TypeVar`, :data:`typing.ParamSpec`,
+  and :data:`typing.TypeVarTuple`) now support defaults.
```
This feels like a very short description of the changes, considering that it involves a change to CPython's grammar 😃 but we can add a more expansive description later; this is non-blocking

AlexWaygood reviewed (approved):
Docs and `typing.py` changes LGTM

hauntsaninja commented on Doc/reference/executionmodel.rst:
```diff
@@ -232,13 +232,17 @@ Annotation scopes differ from function scopes in the following ways:
    Annotation scopes were introduced in Python 3.12 as part of :pep:`695`.
 
+.. versionchanged:: 3.13
+   Annotation scopes are now also used for type parameter defaults, as
```
nit: the "now" makes it sound like annotation scopes previously weren't used for this
```suggestion
   Annotation scopes are also used for type parameter defaults, as
```

hauntsaninja commented on Lib/test/test_unparse.py:
```diff
@@ -746,6 +760,20 @@ def test_async_function_with_type_params(self):
+            ],
+        )
+        ast.fix_missing_locations(node)
+        self.assertEqual(ast.unparse(node), "async def f[T = 1, *Ts = *1, **P = 1]():\n    pass")
```
should we have a `check_ast_roundtrip` case? also maybe good to have it for default values that have a starred expression

JelleZijlstra replied:
Good idea, added

JelleZijlstra committed:
Update Doc/reference/executionmodel.rst

Co-authored-by: Shantanu <12621235+hauntsaninja@users.noreply.github.com>

JelleZijlstra committed:
Roundtrip tests

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pep696v2

JelleZijlstra committed:
Add typing.NoDefault as the default for TypeVar's default

JelleZijlstra committed:
Fix doctest

AlexWaygood commented on Doc/library/ast.rst:
```diff
@@ -1748,43 +1748,55 @@ Type parameters
+   A :class:`typing.TypeVar`. *name* is the name of the type variable.
+   *bound* is the bound or constraints, if any. If *bound* is a :class:`Tuple`,
+   it represents constraints; otherwise it represents the bound. *default_value*
+   is the default value, if any.
```
It might be worth spelling this out a bit more, since this is now different to what happens at runtime. (I think that's the correct choice; the distinction between `None` for no default and `ast.Constant(value=None)` is quite clean in the AST.)

```suggestion
   is the default value; if the :class:`!TypeVar` has no default, this 
   attribute will be set to ``None``.
```

AlexWaygood commented on Doc/library/ast.rst:
```diff
@@ -1748,43 +1748,55 @@ Type parameters
 
-   A :class:`typing.ParamSpec`. ``name`` is the name of the parameter specification.
+   A :class:`typing.ParamSpec`. *name* is the name of the parameter specification.
+   *default_value* is the default value, if any.
```
```suggestion
   *default_value* is the default value; if the :class:`!ParamSpec` has no default,
   this attribute will be set to ``None``.
```

AlexWaygood commented on Doc/library/ast.rst:
```diff
@@ -1796,19 +1808,25 @@ aliases.
 
-   A :class:`typing.TypeVarTuple`. ``name`` is the name of the type variable tuple.
+   A :class:`typing.TypeVarTuple`. *name* is the name of the type variable tuple.
+   *default_value* is the default value, if any.
```
```suggestion
   *default_value* is the default value; if the :class:`!TypeVarTuple` has no
   default, this attribute will be set to ``None``.
```

AlexWaygood commented on Doc/library/typing.rst:
```diff
@@ -1754,15 +1754,27 @@ without the dedicated syntax, as documented below.
 
+   .. attribute:: __default__
+
+      The default value of the type variable, if any. This attribute is ``None``
```
The docs here and for the other `__default__` attributes in `typing.rst` also need to be updated

AlexWaygood commented on Doc/library/typing.rst:
```diff
@@ -3173,6 +3209,13 @@ Introspection helpers
+.. data:: NoDefault
+
+   A sentinel object used to represent the default value of certain parameters
+   used in this module.
```
```suggestion
   A sentinel object used to indicate that a type parameter has no default
   value. For example:
   
   .. doctest::
   
      >>> T = TypeVar("T")
      >>> T.__default__ is typing.NoDefault
      True
      >>> S = TypeVar("S", default=None)
      >>> S.__default__ is None
      True
```

AlexWaygood commented on Doc/reference/compound_stmts.rst:
```diff
@@ -1694,19 +1697,29 @@ evaluated in a separate :ref:`annotation scope <annotation-scopes>`.
+constraints of type variables, the default value is not evaluated when the
+object is created, but only when the type parameter's ``__default__`` attribute
+is accessed. To this end, the default value is evaluated in a separate
+:ref:`annotation scope <annotation-scopes>`.
```
```suggestion
:ref:`annotation scope <annotation-scopes>`. If no default value is specified
for a type parameter, the ``__default__`` attribute is set to the special
sentinel object :data:`typing.NoDefault`.
```

JelleZijlstra committed:
Apply suggestions from code review

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>

JelleZijlstra committed:
Add .has_default() and update docs

AlexWaygood commented on Doc/library/typing.rst:
```diff
@@ -1754,15 +1754,35 @@ without the dedicated syntax, as documented below.
+
+      Return whether or not the type variable has a default value. This is equivalent
+      to checking whether :attr:`__default__` is not equal to :data:`typing.NoDefault`,
+      except that it does not force evaluation of the
```
```suggestion
      to checking whether :attr:`__default__` is not the :data:`typing.NoDefault`
      singleton, except that it does not force evaluation of the
```

AlexWaygood commented on Doc/library/typing.rst:
```diff
@@ -1872,14 +1892,34 @@ without the dedicated syntax, as documented below.
+
+      Return whether or not the type variable tuple has a default value. This is equivalent
+      to checking whether :attr:`__default__` is not equal to :data:`typing.NoDefault`,
+      except that it does not force evaluation of the
```
```suggestion
      to checking whether :attr:`__default__` is not the :data:`typing.NoDefault`
      singleton, except that it does not force evaluation of the
```

AlexWaygood commented on Doc/library/typing.rst:
```diff
@@ -1948,6 +1988,22 @@ without the dedicated syntax, as documented below.
+
+      Return whether or not the parameter specification has a default value. This is equivalent
+      to checking whether :attr:`__default__` is not equal to :data:`typing.NoDefault`,
+      except that it does not force evaluation of the
```
```suggestion
      to checking whether :attr:`__default__` is not the :data:`typing.NoDefault`
      singleton, except that it does not force evaluation of the
```

AlexWaygood reviewed (commented):
I would say "is not the `typing.NoDefault` singleton" rather than "is not equal to `typing.NoDefault`", since we usually argue that an identity check is more idiomatic for singleton objects such as `None`, `NotImplemented` and `Ellipsis`

AlexWaygood commented on Doc/library/typing.rst:
```diff
@@ -3173,6 +3233,22 @@ Introspection helpers
+   .. doctest::
+
+      >>> T = TypeVar("T")
+      >>> T.__default__ is typing.NoDefault
+      True
+      >>> S = TypeVar("S", default=None)
+      >>> S.__default__ is None
+      True
```
```suggestion
      >>> T.has_default()
      False
      >>> T.__default__ is typing.NoDefault
      True
      >>> S = TypeVar("S", default=None)
      >>> S.has_default()
      True
      >>> S.__default__ is None
      True
```

JelleZijlstra replied:
I thought of that but it felt out of place in the docs for `NoDefault`, since the cases you added don't use `NoDefault` at all.

AlexWaygood replied:
Fair enough -- I thought it was quite nice to see together in one example how the two concepts interrelate, but I definitely don't feel strongly!

JelleZijlstra committed:
Apply suggestions from code review

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>

JelleZijlstra committed:
remove trailing space

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pep696v2

JelleZijlstra committed:
Ignore new C globals

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pep696v2

JelleZijlstra committed:
Fix scoping key

The previous logic would add the size of the pointer target to the pointer value, which might point to another pointer being used as a scope key. Now, we increment the value as an integer instead, which means it can never be a valid pointer.
